[docker] Gather more data and expand plugin options

TurboTurtle opened the pull request:
This patch changes the behavior of the docker plugin and collects a bit more
data than before. In addition, it also adds 'docker-engine' to the package lists
as that is the current name of the community release of Docker.

Both 'docker version' and 'docker ps -a' output is now collected. Output from
'docker version' help to validate a given docker binary and 'ps -a' output is
collected to see all terminated, but still existing, containers present on the
system. This is kept separate from normal 'ps' output to keep it simple to only
quickly look at running containers.

Further, plugin options have been changed as follows:

- 'logs' will now capture 'docker logs' output for all running containers.
Previously this was provided by the 'all' option which would also always include
terminated containers.

- 'inspect' will capture 'docker inspect' output for all running containers,
which is helpful in looking at the configuration of a given container.

- 'all' will now cause 'logs' and/or 'inspect' to also include terminated
containers. Note that by itself this option no longer captures any additional
information.

- 'gather_all' will enable all above options. In other words it will capture
'logs' and 'inspect' output for all containers, including terminated ones.

Signed-off-by: Jake Hunsaker <jhunsake@redhat.com>

TurboTurtle commented:
Fixed the pep8 violation.

battlemidget commented:
ack, thanks for the quick response

bmr-cymru commented:
Ditto here; there are fewer options but I do not understand why `logs` and `inspect` are optional at all?

If they can be voluminous on large deployments then I can understand having a mechanism to turn them _off_ but it seems cumbersome to expect users and support associates to remember and apply these myriad options.

TurboTurtle commented:
I can see the case for making inspect enabled by default, but logs I think should definitely be optional. `docker log` pumps out everything sent to standard out by the application so not only is there a volume concern, but there's also the fact that we have no way of reliably obfuscating any data therein.

To keep things consistent I could also see switching the function of "all" and "gather_all".

bmr-cymru commented:
OK: let's enable "inspect" by default and drop that option (unless you think there's a use case for disabling it?)

Ack for the logs, I'd not thought of the scrubbing aspects.

TurboTurtle commented:
Dropped the 'inspect' option and made it just collect inspect output - thinking on it I can't come up with a good reason someone would want to exclude that. Also dropped 'gather_all' since it was basically useless at that point.

codificat commented:
Nice enhancements, thanks!

What about enabling "all" by default? With the default of logs not being collected, hopefully the volume of "inspect" output should be reasonable. I.e. the argument that @bmr-cymru mentioned above: it's cumbersome to have to enable that useful option explicitly when it would likely be useful most of the times: often we're more interested in inspecting containers that have failed.

TurboTurtle commented:
@codificat I fought with that a bit myself. My fear is that putting in on by default would result in a massive dump of ultimately useless data. 

Outside of kubernetes/OpenShift setups where garbage collection is properly configured, most of the docker environments I've seen have **tons** of old, dead containers that are potentially weeks or even months old.

bmr-cymru commented:
The thing to remember when deciding if something should be optional is that Options Are Evil(TM). Users don't discover them or use them when we want and support engineers have to keep on reminding and requesting them.

There are cases where they make sense: normally when the data that would be collected would be burdensome (size) or the time taken to collect would be excessive - the classics here are rotated logs and RPM's `rpm -Va`.

The other common legitimate case is product integration points - e.g. the RHEV/oVirt bits which accept various parameters via plugin option or environment variable that are expected (primarily) to be set by some other program running sos.

These are the things to bear in mind but as the experts for this part of the stack I'll leave the final call with you guys.

sferich888 commented on sos/plugins/docker.py:
```diff
@@ -25,37 +25,52 @@ class Docker(Plugin):
-            "{0} ps".format(self.docker_bin),
-            "{0} images".format(self.docker_bin)
-        ])
+        for cmd in ['info', 'ps', 'ps -a', 'images', 'version']:
```
Should inspect be part of this list? Do we need to have an all or running option like with logs?

TurboTurtle commented:
@sferich888 No, because `inspect` requires a container ID which is why it is run later. The list referenced is collecting info from the daemon "itself", not containers specifically.

sferich888 commented:
Ah missed it later in the file

TurboTurtle commented:
@codificat coming back around to this one - I'm still in the "don't enable all by default" camp. I really, really want to avoid having a massive dump of dozens or perhaps hundreds of `inspect` files in every sosreport. Any further thoughts?

@bmr-cymru Assuming the above gets sorted, anything else this needs in your eyes to get pulled in?

sferich888 commented:
@TurboTurtle I get not wanting to collect a lot of files, however openshift and kubernetties performance profiling limits a node to ~40 -> ~120 pods per node (assuming the node has the resources to run the pods). 

That said the size of docker inspect output (from the man page) is ~4.5kb. This translates to, from a size perspective, ~500kb (on the high end) of data we collet by running this. This could grow over time, but even at these values nodes have to have 60+ GB of RAM (assuming 512 MB per pod).

It's not a lot compared to what other products collect, however the real question with this is do we want to collect this in the background and have it, or should we ask for this when we need it is specific cases?

Another way to think about this is, is there any data in this output that we could use to proactivly solve issues? If so I think the 500kb is worth grabbing this data.

codificat commented:
@sferich888 the plugin will run on non-ose/k8s systems too. Even on k8s systems, the limit is on _running_ pods, there will likely be non-running containers that would be inspect'd. Also, a pod can contain multiple containers. Plus, I only checked an `inspect` on a default ose infra container and it's 8+ kb. In short: I understand @TurboTurtle 's concern about the amount of data collected.
Moreover, `inspect` output is also susceptible to expose sensitive data in environment variables and we should implement something along the lines of what's being discussed in #718 / #780.
Having said that: that output is indeed valuable...

TurboTurtle commented:
@codificat @sferich888

Ok, so sat down with the entire container support team over here and the agreement is unanimously that the `all` option should **not** be enabled by default, so that's what I'm going with. 

@bmr-cymru Made the change to `docker_cmd` from the discussion in #718 and rebased for `add_journal`. At this point are there any blockers to merging this?

battlemidget commented:
@TurboTurtle ill ping @bmr-cymru again to see about this PR

pmoravec commented:
+1 for the latest commit

battlemidget commented:
This looks to have been merged already
